### PR TITLE
e2e: fix eventual consistency failure within consultemplate suite.

### PR DESCRIPTION
**Pull request by jrasell**
Failure: https://app.circleci.com/pipelines/github/hashicorp/nomad-e2e/472/workflows/0a855d95-b76d-4f88-bd53-a35082b7cce9/jobs/2717

Tested with a full E2E cluster:
```
=== RUN   TestE2E
=== RUN   TestE2E/ConsulTemplate
=== RUN   TestE2E/ConsulTemplate/*consultemplate.ConsulTemplateTest
=== RUN   TestE2E/ConsulTemplate/*consultemplate.ConsulTemplateTest/TestConsulTemplate_NomadServiceLookups
--- PASS: TestE2E (18.95s)
    --- PASS: TestE2E/ConsulTemplate (18.95s)
        --- PASS: TestE2E/ConsulTemplate/*consultemplate.ConsulTemplateTest (18.94s)
            --- PASS: TestE2E/ConsulTemplate/*consultemplate.ConsulTemplateTest/TestConsulTemplate_NomadServiceLookups (18.42s)
PASS
ok  	github.com/hashicorp/nomad/e2e	19.487s
```

### CODE REVIEW

File: e2e/consultemplate/consultemplate.go [hunk 1]
```diff
@@ -447,10 +447,18 @@ func (tc *ConsulTemplateTest) TestConsulTemplate_NomadServiceLookups(f *framewor
...
 	// Pull the allocation ID for the job, we use this to ensure this is found
-	// in the rendered template later on.
-	serviceJobAllocs, err = e2eutil.AllocsForJob(serviceJobID, "default")
-	f.NoError(err)
-	f.Len(serviceJobAllocs, 3)
+	// in the rendered template later on. Wrap this in an eventual do to the
+	// eventual consistency around the service registration process.
+	f.Eventually(func() bool {
+		serviceJobAllocs, err = e2eutil.AllocsForJob(serviceJobID, "default")
+		if !f.Assert().NoError(err) {
```
Comment (tgross) [inline e2e/consultemplate/consultemplate.go:454]: Checking the return value here lets us return false to `Eventually`, but doesn't it fail the surrounding test even if it passes later on?

Comment (jrasell) [reply e2e/consultemplate/consultemplate.go:454]: I'm not sure I follow, so am keen to understand what I am missing here.

As I understand it, each "tick" of the eventually function would run the enclosed functions until true is returned or the timeout is reached. In the case `e2eutil.AllocsForJob` returns an error, the eventually function will tick again, the same if the length assertion fails; however, if both are found as we expect, true is returned and the eventually trigger exist successfully.

Comment (tgross) [reply e2e/consultemplate/consultemplate.go:454]: Right, but assert also sets the "failed" state on the overall test. So suppose you have the following test:

```go
package e2eutil

import (
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestWhatever(t *testing.T) {
	i := 0
	require.Eventually(t, func() bool {
		i++
		if !assert.Greater(t, i, 3) {
			return false
		}
		return true
	}, time.Second, time.Millisecond*10)

}
```

You end up with 3 test assertion failures, even though the `Eventually` test assertion passes:

```
$ go test -v . -run TestWhatever
=== RUN   TestWhatever
    example_test.go:15:
                Error Trace:    example_test.go:15
                                                        asm_amd64.s:1581
                Error:          "1" is not greater than "3"
                Test:           TestWhatever
    example_test.go:15:
                Error Trace:    example_test.go:15
                                                        asm_amd64.s:1581
                Error:          "2" is not greater than "3"
                Test:           TestWhatever
    example_test.go:15:
                Error Trace:    example_test.go:15
                                                        asm_amd64.s:1581
                Error:          "3" is not greater than "3"
                Test:           TestWhatever
--- FAIL: TestWhatever (0.04s)
FAIL
FAIL    github.com/hashicorp/nomad/e2e/e2eutil  1.032s
FAIL
```

Compare this to:

```go
func TestWhateverImproved(t *testing.T) {
	i := 0
	require.Eventually(t, func() bool {
		i++
		if i < 3 {
			return false
		}
		return true
	}, time.Second, time.Millisecond*10)
}
```

Which passes:

```
$ go test -v . -run TestWhateverImproved
=== RUN   TestWhateverImproved
--- PASS: TestWhateverImproved (0.03s)
PASS
ok      github.com/hashicorp/nomad/e2e/e2eutil  0.931s
```

Comment (jrasell) [reply e2e/consultemplate/consultemplate.go:454]: Hot damn, I didn't know that but now it seems so obvious. Thanks a lot @tgross! I'll get this fixed-up and pushed.

**Pull request merged**